h2o-core/src/main/java/hex/BulkModelBuilder.java

```diff
@@ -0,0 +1,86 @@
+package hex;
+
+import water.H2O;
+import water.Job;
+import water.util.Log;
+
+/**
+ * Runs given model builders in bulk.
+ */
+public class BulkModelBuilder {
+
+    private final String modelType;
+    private final Job job;
+    private final ModelBuilder<?, ?, ?>[] modelBuilders;
+    private final int parallelization;
+    private final int updateInc;
+
+    /**
+     * @param modelType       text description of group of models being built (for logging purposes)
+     * @param job             parent job (processing will be stopped if stop of a parent job was requested)
+     * @param modelBuilders   list of model builders to run in bulk
+     * @param parallelization level of parallelization (how many models can be built at the same time)
+     * @param updateInc       update increment (0 = disable updates)
+     */
+    public BulkModelBuilder(
+        String modelType, Job job, ModelBuilder<?, ?, ?>[] modelBuilders,
+        int parallelization, int updateInc
+    ) {
+        this.modelType = modelType;
+        this.job = job;
+        this.modelBuilders = modelBuilders;
+        this.parallelization = parallelization;
+        this.updateInc = updateInc;
+    }
+    
+    protected void prepare(ModelBuilder m) {}
+    
+    protected void finished(ModelBuilder m) {}
+
+    public void bulkBuildModels() {
+        final int N = modelBuilders.length;
+        H2O.H2OCountedCompleter[] submodel_tasks = new H2O.H2OCountedCompleter[N];
+        int nRunning = 0;
+        RuntimeException rt = null;
+        for (int i = 0; i < N; ++i) {
+            if (job.stop_requested()) {
+                Log.info("Skipping build of last " + (N - i) + " out of " + N + " " + modelType + " CV models");
+                stopAll(submodel_tasks);
+                throw new Job.JobCancelledException();
+            }
+            Log.info("Building " + modelType + " model " + (i + 1) + " / " + N + ".");
+            prepare(modelBuilders[i]);
+            submodel_tasks[i] = H2O.submitTask(modelBuilders[i].trainModelImpl());
+            if (++nRunning == parallelization) { //piece-wise advance in training the models
+                while (nRunning > 0) try {
+                    int waitForTaskIndex = i + 1 - nRunning;
+                    submodel_tasks[waitForTaskIndex].join();
```

Comment:
This is the reason why this method is not efficient in the general case of model building. If the model you are waiting for happens to take a much longer time then the rest then you will be underutilizing you resources: you have one you are waiting for and n-1 already completed. In the case of GPUs it is especially bad because these are not a shared resource as compared to CPUs.

I would rename this class to CVModelBuilder or such and/or modify the javadoc.

Reply:
Take a look at ModelBuilderHelper#trainModelsParallel for a generic and efficient parallel model builder.

It should be somewhat clear when to use "BulkModelBuilder" and when `trainModelsParallel`. Maybe ModelBuilderHelper needs to be refactored? And we could have CVModelBuilder and ParallelModelBuilder?

WDYT?

Reply:
improved things a bit, bigger refactor might be in order